### PR TITLE
Hw2

### PR DESCRIPTION
**Pull request by mikiyas17legesse**
# Pull Request

## Summary
- What problem is being solved?
- Why now? Any user-visible impact?

## Related Issues
- Closes #
- Relates to #

## Change Type
- [ ] 🐛 Bug fix
- [ ] ✨ Feature / enhancement
- [ ] 💥 Breaking change
- [ ] 📚 Documentation
- [ ] 🔧 Refactor / cleanup
- [ ] ⚡ Performance
- [ ] 🧪 Test improvement
- [ ] 🔨 Build / tooling

## Impacted Areas
- [ ] `mail_client_api`
- [ ] `gmail_client_impl`
- [ ] Documentation
- [ ] Tests
- [ ] Tooling / CI
- [ ] Other (describe in summary)

### Commands executed
```bash
# Example (edit as needed)
uv run pytest src/ tests/
uv run mypy src tests
uv run ruff check .
```

### Results
- [ ] Tests pass
- [ ] Type checks clean
- [ ] Linting clean
- [ ] Manual verification (if applicable)

## Quality Checklist
- [ ] Contract boundaries respected; abstractions unchanged unless noted
- [ ] Backward compatibility confirmed or migration documented
- [ ] Security / secrets handled correctly
- [ ] Performance impact acceptable
- [ ] Docs updated for user-facing changes

## Breaking Changes (skip if none)
1. Impact:
2. Migration / mitigation:

## Notes for Reviewers
Call out tricky areas, follow-ups, or validation details.

### CODE REVIEW

File: jira_tokens.db
Comment (austinhuang0131) [inline jira_tokens.db:1]: Do not commit the production database!!! I can literally extract the OAuth token from it.

File: docs/architecture.md [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+- **OAuth 2.0 Integration**: Secure authentication with token management and refresh
+
+## System Overview
+
+```
+┌─────────────────────────────────────────────────────────────┐
+│                    Client Applications                       │
+│                 (Web, Mobile, CLI, etc.)                   │
+└─────────────────────────┬───────────────────────────────────┘
+                          │ HTTP/REST
```
Comment (austinhuang0131) [inline docs/architecture.md]:
```suggestion
                          │ 
```
There is no HTTP/REST here.

File: docs/architecture.md [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+│                  ticket_service                             │
+│              (FastAPI HTTP Service)                         │
+│  • REST API endpoints                                       │
+│  • Request/Response validation                              │
+│  • OAuth 2.0 authentication                                │
+│  • OpenAPI documentation                                    │
+└─────────────────────────┬───────────────────────────────────┘
+                          │ TicketServiceAPI
+┌─────────────────────────▼───────────────────────────────────┐
+│                   ticket_impl                               │
```
Comment (austinhuang0131) [inline docs/architecture.md]: A client can choose to use `ticket_impl` directly, so technically there should be another arrow from client to it.

File: src/ticket_api/tests/test_edge_cases.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+EXPECTED_COMMENT_COUNT = 3
+
+
+class TestEdgeCases:
+    """Test edge cases and boundary conditions."""
+
+    def test_ticket_with_unicode_content(self) -> None:
+        """Test ticket with unicode characters."""
+        ticket = Ticket(
+            title="Bug with émojis",
```
Comment (austinhuang0131) [inline src/ticket_api/tests/test_edge_cases.py:23]: That's not an emoji. It is also dubious whether it's testing code that you wrote (which is what unit tests should be) or testing some underlying library or even just the language's features, so whether the test is even necessary is a question (it would be if, say, you forbid non-ASCII characters).

File: src/ticket_impl/src/ticket_impl/impl.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+        author=(c.get("author") or {}).get("displayName") or "unknown",
+        content=(text.strip() or "<empty>"),
+    )
+
+
+# ---- concrete implementation ----
+
+
+class TicketImpl(TicketServiceAPI):
+    """Jira-backed implementation of the TicketServiceAPI. All methods are async and map domain UUIDs to Jira issue keys internally."""  # noqa: E501
```
Comment (austinhuang0131) [inline src/ticket_impl/src/ticket_impl/impl.py]: Unnecessary `noqa`, just break the docstring

File: src/ticket_impl/src/ticket_impl/storage.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
+"""Persistence helpers: OAuth tokens and (uuid ↔ Jira key) mapping."""
```
Comment (austinhuang0131) [inline src/ticket_impl/src/ticket_impl/storage.py:1]: FYI they said no database and you're supposed to use session tokens or something. Though I suppose none of the teams realized this before they added it to the assignment description...

File: .gitignore [hunk 6]
```diff
@@ -31,6 +31,7 @@ site/
 token.json
 .env
 credentials.json
+jira_tokens.db
```
Comment (austinhuang0131) [inline .gitignore:34]: You still have to remove the file from the repository if you have already committed it.

File: tests/e2e/test_e2e_workflows.py [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+
+
+class TestE2ETicketManagement:
+    """End-to-end tests for complete ticket management workflows."""
+
+    def test_e2e_team_collaboration_workflow(self, mock_jira_backend: dict[str, MagicMock]) -> None:
+        """Test team collaboration workflow.
+
+        E2E test simulating a real team collaboration scenario:
+        1. Manager creates a ticket
```
Comment (austinhuang0131) [inline tests/e2e/test_e2e_workflows.py]: You should probably start from OAuth in a e2e test.

Also it is my understanding that e2e tests should test using real credentials against the actual API; if you're still mocking some parts and testing others, then it's an integration test.

File: tests/e2e/test_e2e_workflows.py [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+            headers={
+                "X-User-ID": "test-developer",
+                "X-Project-Key": "BUG",
+            },
+        )
+        assert response.status_code == HTTP_OK
+
+
+class TestE2EDataConsistency:
+    """E2E tests for data consistency across operations."""
```
Comment (austinhuang0131) [inline tests/e2e/test_e2e_workflows.py]: Does it matter to test whether the data is consistent? It's handled by Jira, not by your code... (Besides in a real-world scenario your client is not the only one making changes, making the test moot.)

File: tests/integration/test_service_adapter_integration.py [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+        assert tickets[0]["title"] == "Fix login bug"
+        assert tickets[1]["title"] == "Database optimization"
+
+    def test_add_comment_through_service(
+        self,
+        sample_ticket: Ticket,
+        mock_jira_backend: dict[str, MagicMock],
+    ) -> None:
+        """Test adding a comment through the REST API."""
+        from ticket_api import Comment
```
Comment (austinhuang0131) [inline tests/integration/test_service_adapter_integration.py]: Probably should be imported on the top

File: src/ticket_impl/src/ticket_impl/oauth.py [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+
+logger = logging.getLogger(__name__)
+
+AUTH_BASE = "https://auth.atlassian.com"
+TOKEN_URL = f"{AUTH_BASE}/oauth/token"
+SCOPE = "read:jira-user read:jira-work write:jira-work offline_access"
+AUDIENCE = "api.atlassian.com"
+
+# HTTP status codes
+HTTP_OK = 200
```
Comment (austinhuang0131) [inline src/ticket_impl/src/ticket_impl/oauth.py]:
```python
from http import HTTPStatus
```

**Review by austinhuang0131 — COMMENTED** [review]
Overall I'd say it's according to specification (save the database part).

Also you shouldn't need to remove the old gmail parts, it mixes with the PR changes and makes it slightly harder to review.

File: src/ticket_client_adapter/pyproject.toml [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+description = "Service adapter wrapping auto-generated client with TicketServiceAPI"
+readme = "README.md"
+requires-python = ">=3.11"
+dependencies = [
+    "ticket-api",
+    "ticket-service-client",  # The auto-generated client!
+]
+
+[tool.hatch.build.targets.wheel]
+packages = ["src/ticket_client_adapter"]
```
Comment (austinhuang0131) [inline src/ticket_client_adapter/pyproject.toml:17]: Missing ruff config extend

File: src/ticket_service/tests/test_endpoints.py [hunk 12]
```diff
@@ -0,0 +1,644 @@
...
+        sample_tickets: list[Ticket],
+    ) -> None:
+        """Test listing tickets with status filter."""
+        with patch("ticket_service.main.TicketImpl") as mock_impl:
+            mock_service = AsyncMock()
+            mock_service.list_tickets.return_value = [sample_tickets[0]]
+            mock_impl.return_value = mock_service
+
+            response = await async_client.get(
+                "/api/v1/tickets?status=open&assignee=assignee@example.com",
```
Comment (austinhuang0131) [inline src/ticket_service/tests/test_endpoints.py]: Use `params`.

File: jira_tokens.db
Comment (ivanearisty) [reply jira_tokens.db:1]: +1

File: src/ticket_api/pyproject.toml [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+build-backend = "hatchling.build"
+
+[project]
+name = "ticket-api"
+version = "0.1.0"
+description = "Abstract interface and data models for ticketing microservice"
+readme = "README.md"
+requires-python = ">=3.11"
+dependencies = [
+    "pydantic",
```
Comment (ivanearisty) [inline src/ticket_api/pyproject.toml]: We were using built ins like ABC before. Pydantic should not be used here.

File: src/ticket_api/src/ticket_api/__init__.py [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+
+This package defines the base contract for ticket operations including:
+- Data models (Ticket, Comment)
+- Abstract service interface (TicketServiceAPI)
+"""
+
+from .interface import TicketServiceAPI
+from .models import Comment, Ticket, TicketPriority, TicketStatus
+
+__all__ = ["Comment", "Ticket", "TicketPriority", "TicketServiceAPI", "TicketStatus"]
```
Comment (ivanearisty) [inline src/ticket_api/src/ticket_api/__init__.py]: We're not using this all paradigm anymore.

It's not really necessary and just a thing all AIs add.

__all__ only controls wildcard imports (from pkg import *). 

If you don’t use wildcard imports, you don’t need it.

File: src/ticket_api/src/ticket_api/interface.py [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+must follow, ensuring consistency across different implementations.
+"""
+
+from abc import ABC, abstractmethod
+from uuid import UUID
+
+from .models import Comment, Ticket, TicketPriority, TicketStatus
+
+
+class TicketServiceAPI(ABC):
```
Comment (ivanearisty) [inline src/ticket_api/src/ticket_api/interface.py:13]: Wait so this is an interface using ABCs but we still have pydantic?

File: src/ticket_api/src/ticket_api/interface.py
Comment (ivanearisty) [inline src/ticket_api/src/ticket_api/interface.py:1]: Interface methods look good to me

File: src/ticket_api/src/ticket_api/models.py
Comment (ivanearisty) [inline src/ticket_api/src/ticket_api/models.py:1]: This entire class is wrong. 

It contains business logic methods (add_comment, update_status, assign_to). 

Business actions should be on the port/service, not the DTO.

File: src/ticket_api/src/ticket_api/models.py [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+    content: str = Field(min_length=1, max_length=2000, description="Comment text content")
+    created_at: datetime = Field(
+        default_factory=lambda: datetime.now(UTC),
+        description="When the comment was created",
+    )
+
+    model_config = ConfigDict(frozen=True)  # Make comments immutable
+
+
+class Ticket(BaseModel):
```
Comment (ivanearisty) [inline src/ticket_api/src/ticket_api/models.py]: @dataclass(frozen=True)
class Ticket:
    id: UUID
    title: str
    (other stuff)

File: src/ticket_api/src/ticket_api/models.py [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+    updated_at: datetime = Field(
+        default_factory=lambda: datetime.now(UTC),
+        description="When the ticket was last updated",
+    )
+    comments: list[Comment] = Field(
+        default_factory=list,
+        description="List of comments on this ticket",
+    )
+
+    def add_comment(self, author: str, content: str) -> "Ticket":
```
Comment (ivanearisty) [inline src/ticket_api/src/ticket_api/models.py]: tickets should not be responsible for their own state

File: src/ticket_api/src/ticket_api/models.py [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+class TicketPriority(str, Enum):
+    """Enumeration of ticket priority levels."""
+
+    LOW = "low"
+    MEDIUM = "medium"
+    HIGH = "high"
+    CRITICAL = "critical"
+
+
+class Comment(BaseModel):
```
Comment (ivanearisty) [inline src/ticket_api/src/ticket_api/models.py]: This is also wrong, this should be at the network edge for data validation.

File: src/ticket_api/tests/test_api_contract.py
Comment (ivanearisty) [inline src/ticket_api/tests/test_api_contract.py:1]: These should change to reflect that tickets don't update themselves

File: src/ticket_api/src/ticket_api/interface.py [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+            List of Ticket instances matching the criteria
+
+        Raises:
+            ValueError: If limit is negative or offset is negative
+            ServiceError: If listing operation fails
+
+        """
+
+    @abstractmethod
+    async def update_ticket(  # noqa: PLR0913
```
Comment (ivanearisty) [inline src/ticket_api/src/ticket_api/interface.py]: more explicit operations like transition_status or reassign_ticket could better model domain events

File: src/ticket_api/pyproject.toml
Comment (ivanearisty) [inline src/ticket_api/pyproject.toml:1]: Instead of conditionally returning None, just throw a custom error.

File: src/ticket_client_adapter/src/ticket_client_adapter/client.py
Comment (ivanearisty) [inline src/ticket_client_adapter/src/ticket_client_adapter/client.py:1]: There is no documentation or visible implementation of idempotency strategies (e.g., using an Idempotency-Key header) or retry logic (e.g., with exponential backoff) in the adapters.

File: src/ticket_impl/pyproject.toml [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+requires = ["hatchling"]
+build-backend = "hatchling.build"
+
+[project]
+name = "ticket-impl"
+version = "0.1.0"
+description = "Jira-backed implementation of ticket_api"
+readme = "README.md"
+requires-python = ">=3.11"
+# IMPORTANT: set this to the exact project name of your API package
```
Comment (ivanearisty) [inline src/ticket_impl/pyproject.toml]: remove these comments

File: src/ticket_impl/src/ticket_impl/oauth.py [hunk 21]
```diff
@@ -0,0 +1,233 @@
...
+        r = await client.post(TOKEN_URL, json=data)
+        r.raise_for_status()
+        payload = r.json()
+    access = cast("str", payload["access_token"])
+    expires_in = int(payload.get("expires_in", 3600))
+    update_access(user_id, access, expires_in)
+    return access
+
+
+async def get_valid_access_token(user_id: str) -> str:
```
Comment (ivanearisty) [inline src/ticket_impl/src/ticket_impl/oauth.py:131]: this is fine sure

File: src/ticket_impl/src/ticket_impl/storage.py (hunk 5 above)
Comment (ivanearisty) [reply src/ticket_impl/src/ticket_impl/storage.py:1]: Updating this comment based on some of Nikolai's feedback:

Several students have implemented a small db to store OAuth credentials.

Since this is an in-memory database, once you have more than one instance running, it will no longer work.

If instead of doing a DB, you had a simple global dict that kept the state of all users, it'd effectively be the same thing as having an in-memory data store. You could persist the DB to JSON on shutdown/update as well.

So, the only real differences between having a SQLlite DB or just using a plain dict are not supporting multiple instances and some security/hashing concerns.

TL;DR you'd need a separate place were info is kept, agnostic of the FastAPI service, to actually be able to support multiple instances.

For this HW, I'm pretty sure that the in-memory store is fine. We'll probably switch things up in a later HW. I'll update later.

File: tests/e2e/test_e2e_workflows.py [hunk 22]
```diff
@@ -0,0 +1,399 @@
...
+            assignee="developer@example.com",
+            created_at=datetime.fromisoformat("2024-01-20T09:00:00+00:00"),
+            updated_at=datetime.fromisoformat("2024-01-20T16:00:00+00:00"),
+            comments=[comment1, comment2],
+        )
+
+        client = TestClient(app)
+
+        # Step 1: Manager creates ticket
+        mock_jira_backend["create_ticket"].return_value = created
```
Comment (ivanearisty) [inline tests/e2e/test_e2e_workflows.py]: Why are we mocking jira in an e2e test?

File: tests/e2e/test_e2e_workflows.py [hunk 23]
```diff
@@ -0,0 +1,399 @@
...
+                "priority": "high",
+            },
+            headers={
+                "X-User-ID": "test-team-lead",
+                "X-Project-Key": "TASK",
+            },
+        )
+        assert response.status_code == HTTP_CREATED
+
+        # Simulate getting ticket multiple times
```
Comment (ivanearisty) [inline tests/e2e/test_e2e_workflows.py]: Simulating in an e2e test. Why?

File: tests/e2e/test_e2e_workflows.py [hunk 24]
```diff
@@ -0,0 +1,399 @@
...
+        mock_jira_backend["get_ticket"].return_value = ticket
+        for _ in range(3):
+            response = client.get(
```
Comment (ivanearisty) [inline tests/e2e/test_e2e_workflows.py]: wait so are we actually calling the jira backend or?

**Review by ivanearisty — CHANGES_REQUESTED** [review]

**Comment by ivanearisty** [comment]
I was thinking about ABC vs @dataclass for your interfaces. Use ABC.

ABC = “contract/interface.” Use when you want polymorphism (many backends implement the same API) and you’ll call code through that interface.
@dataclass = “value object.” Use when you just need to hold data with minimal boilerplate (init/eq/repr), possibly immutable.

ABC is better in our case:
- It's better for multiple implementations (e.g., GmailMessage, OutlookMessage, IMAPMessage).
- You want to enforce required attributes/methods regardless of storage/transport.
- You need behavior (methods) beyond plain data.
- You care about LSP-friendly substitution and unit-testable fakes.

Hence, for things like a message, do as the gmail example for contracts like:
```python3
class Message(ABC):
    """Abstract base class representing an email message."""
    @property
    @abstractmethod
    def id(self) -> str:
        """Return the unique identifier of the message."""
        raise NotImplementedError
```

I'm leaving this on every repo btw so it might or might not apply to you. But I think everyone has at least one issue regarding this.
